File: backend/video_processor.py
import os
import subprocess


BASE_DIR = os.path.dirname(os.path.abspath(__file__))
CLIPS_DIR = os.path.join(BASE_DIR, "clips")
os.makedirs(CLIPS_DIR, exist_ok=True)


def generate_thumbnail(video_path: str, clip_id: str) -> str:
    thumb_path = os.path.join(CLIPS_DIR, f"{clip_id}_thumb.jpg")
    result = subprocess.run(
        [
            "ffmpeg",
            "-y",
            "-ss",
            "00:00:01",
            "-i",
            video_path,
            "-frames:v",
            "1",
            "-q:v",
            "2",
            thumb_path,
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(result.stderr or "Thumbnail generation failed")
    return thumb_path

File: backend/video_processor.js
const fs = require('fs')
const path = require('path')
const { execFile } = require('child_process')

const BASE_DIR = __dirname
const STORAGE_DIR = process.env.STORAGE_DIR || BASE_DIR
const CLIPS_DIR = path.join(STORAGE_DIR, 'clips')
fs.mkdirSync(CLIPS_DIR, { recursive: true })

function generateThumbnail(videoPath, clipId) {
  const thumbPath = path.join(CLIPS_DIR, `${clipId}_thumb.jpg`)
  return new Promise((resolve, reject) => {
    execFile('ffmpeg', [
      '-y',
      '-ss', '00:00:01',
      '-i', videoPath,
      '-frames:v', '1',
      '-q:v', '2',
      thumbPath,
    ], { maxBuffer: 16 * 1024 * 1024, timeout: 30000 }, (error, stdout, stderr) => {
      if (error) {
        reject(new Error(stderr || 'Thumbnail generation failed'))
        return
      }
      resolve(thumbPath)
    })
  })
}

module.exports = { generateThumbnail }
